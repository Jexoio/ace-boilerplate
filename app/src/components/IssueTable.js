// @flow
/* eslint-disable camelcase */
import React, { Component } from 'react';
import InlineEdit, { SingleLineTextInput } from '@atlaskit/inline-edit';
import styled from 'styled-components';
import Theme from '../providers/Theme';

type IssueType = {
  summary: string,
  id: string,
  key: string
};
type IssueProps = {
  issue: IssueType,
  updateSummary: Function
};
type IssueState = {
  summary: string,
  previousSummary: string
};
type IssuesTableProps = {
  updateSummary: Function,
  subscribeToJiraIssueUpdate: Function,
  issues: Array<IssueType>
};

const Panel = styled.div`
  h1 {
    position: relative;
    border-radius: 6px 6px 0 0;
    border: 1px solid transparent;
    background-color: ${({ theme }) => `${theme.panel.header.backgroundColor}`};
    color: ${({ theme }) => `${theme.panel.header.textColor}`};
    padding: 16px;
    margin: 0px;
  }
  .body {
    padding: 16px;
    border: ${({ theme }) => `1px solid ${theme.panel.body.borderColor}`};
  }
`;

const IssueRow = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-content: flex-start;
  height: 40px;
  > * {
    flex: 1;
    line-height: 40px;
  }
  .summary {
    flex: 2;
  }
`;

export class Issue extends Component<IssueProps, IssueState> {
  state = {
    summary: this.props.issue.summary,
    previousSummary: this.props.issue.summary,
  };

  UNSAFE_componentWillUpdate({ issue }: IssueProps, { summary }: IssueState) {
    if (issue.summary !== summary && summary === this.state.summary) {
      this.setState({ summary: issue.summary, previousSummary: issue.summary });
    }
  }

  handleChange = (e: any) => {
    const summary = e.target.value;
    this.setState({ summary });
  };

  handleConfirm = () => {
    this.props.updateSummary({
      variables: { id: this.props.issue.id, summary: this.state.summary },
    });
  };

  handleCancel = () => {
    this.setState({ summary: this.state.previousSummary });
  };

  render() {
    const { issue } = this.props;
    return (
      <IssueRow key={issue.id}>
        <div>{issue.id}</div>
        <div>{issue.key}</div>
        <div className="summary">
          <InlineEdit
            isFitContainerWidthReadView
            isLabelHidden
            editView={
              <SingleLineTextInput
                id="1"
                isEditing
                isInitiallySelected
                value={this.state.summary}
                onChange={this.handleChange}
              />
            }
            readView={
              <SingleLineTextInput
                id="1"
                isEditing={false}
                isInitiallySelected
                value={this.state.summary}
                onChange={this.handleChange}
              />
            }
            onConfirm={this.handleConfirm}
            onCancel={this.handleCancel}
          />
        </div>
      </IssueRow>
    );
  }
}

class IssuesTable extends Component<IssuesTableProps> {
  unsubscribe: Function;

  componentDidMount() {
    this.unsubscribe = this.props.subscribeToJiraIssueUpdate();
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  render() {
    const { issues } = this.props;
    return (
      <Theme.Consumer>
        {tokens => (
          <Panel theme={tokens}>
            <h1>My latest issues</h1>
            <div className="body">
              {issues.map(issue => (
                <Issue
                  issue={issue}
                  key={issue.id}
                  updateSummary={this.props.updateSummary}
                />
              ))}
            </div>
          </Panel>
        )}
      </Theme.Consumer>
    );
  }
}

export default IssuesTable;
