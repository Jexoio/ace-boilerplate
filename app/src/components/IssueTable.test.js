import React from 'react';
import { shallow, wrapper } from 'enzyme';
import renderer from 'react-test-renderer';
import IssueTable, { Issue } from './IssueTable';

const issue = {
  summary: 'foo',
  id: 'bar',
  key: 'baz'
};
describe('<Issue />', () => {
  it('renders without crashing', () => {
    shallow(<Issue issue={issue} />);
  });
  it('matches the snapshot', () => {
    const tree = renderer.create(<Issue issue={issue} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  describe('handleChange', () => {
    it('updates the state', () => {
      const wrapper = shallow(<Issue issue={issue} />);
      wrapper.instance().handleChange({ target: { value: 'new value' } });
      expect(wrapper.instance().state.summary).toBe('new value');
    });
  });
  describe('handleConfirm', () => {
    it('updates the state', () => {
      const updateSummary = jest.fn();
      const wrapper = shallow(
        <Issue issue={issue} updateSummary={updateSummary} />
      );
      wrapper.instance().handleConfirm({ target: { value: 'new value' } });
      expect(updateSummary.mock.calls.length).toBe(1);
    });
  });
  describe('handleCancel', () => {
    it('updates the state', () => {
      const updateSummary = jest.fn();
      const wrapper = shallow(<Issue issue={issue} />);
      const instance = wrapper.instance();
      instance.handleChange({ target: { value: 'new value' } });
      instance.handleCancel();
      expect(wrapper.instance().state.summary).toBe(issue.summary);
    });
  });
});

describe('<IssueTable />', () => {
  it('renders without crashing', () => {
    shallow(<IssueTable subscribeToJiraIssueUpdate={() => ({})} />);
  });
  it('matches the snapshot', () => {
    const tree = renderer
      .create(
        <IssueTable subscribeToJiraIssueUpdate={() => ({})} issues={[]} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('subscribes/unsubscribes to jira issue updates', () => {
    const unsubscribe = jest.fn();
    const subscribe = jest.fn().mockReturnValue(unsubscribe);
    const component = shallow(
      <IssueTable subscribeToJiraIssueUpdate={subscribe} />
    );
    expect(subscribe.mock.calls.length).toBe(1);
    component.unmount();
    expect(unsubscribe.mock.calls.length).toBe(1);
  });
});
