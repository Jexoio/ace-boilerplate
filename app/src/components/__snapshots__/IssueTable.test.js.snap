// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Issue /> matches the snapshot 1`] = `
<div
  className="sc-gzVnrw emaAvU"
>
  <div>
    bar
  </div>
  <div>
    baz
  </div>
  <div
    className="summary"
  >
    <div
      className="RootWrapper-sc-1va80k6-0 fsVJT"
    >
      <div
        style={
          Object {
            "position": "absolute",
          }
        }
      >
        <label
          className="Label__LabelWrapper-sc-17towfw-0 keOXhT"
        >
          <div
            className="Label__LabelInner-sc-17towfw-1 jbyPaz"
          >
            <span
              onClick={[Function]}
            >
              
            </span>
          </div>
        </label>
      </div>
      <div
        className="ContentWrapper-kdagst-0 fPLsyX"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <div
          className="FieldBaseWrapper-sc-14kmybr-0 cspqsP"
          onClick={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="Content__ContentWrapper-ve26fj-2 feWTNh"
            disabled={false}
          >
            <div
              className="Content__ChildWrapper-ve26fj-0 ilJUcG"
            >
              <div
                className="Content-ve26fj-1 llCsQv"
                disabled={false}
                onBlurCapture={[Function]}
                onFocusCapture={[Function]}
                readOnly={false}
              >
                <div
                  className="ReadViewContentWrapper-xymwx5-0 kLiHRY"
                >
                  <div
                    className="Input__ReadView-sc-1b6ayio-0 bZECeW"
                    style={Object {}}
                  >
                    foo
                  </div>
                  <button
                    aria-label="Edit"
                    className="EditButton-sc-1v6bv8a-0 doKTJP"
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<IssueTable /> matches the snapshot 1`] = `
<div
  className="sc-bZQynM iYOxHk"
>
  <h1>
    My latest issues
  </h1>
  <div
    className="body"
  />
</div>
`;
